src/NServiceBus.Core/Forwarding/InvokeForwardingPipelineBehavior.cs

```diff
@@ -0,0 +1,45 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using NServiceBus.Forwarding;
+    using NServiceBus.Pipeline;
+    using NServiceBus.Routing;
+    using NServiceBus.Transports;
+
+    class InvokeForwardingPipelineBehavior : PhysicalMessageProcessingStageBehavior
+    {
+        public InvokeForwardingPipelineBehavior(IPipelineBase<ForwardingContext> forwardingPipeline, string forwardingAddress)
+        {
+            this.forwardingPipeline = forwardingPipeline;
+            this.forwardingAddress = forwardingAddress;
+        }
+
+        public override void Invoke(Context context, Action next)
+        {
+            next();
+
+            context.PhysicalMessage.RevertToOriginalBodyIfNeeded();
+
+            var processedMessage = new OutgoingMessage(context.PhysicalMessage.Id, context.PhysicalMessage.Headers, context.PhysicalMessage.Body);
+
+            var forwardingContext = new ForwardingContext(processedMessage,context);
```

Comment:
After talking it through with @andreasohlund I think we should keep the context constructors minimal (as here) to be able to instantiate these contexts more easily for testing purposes (cc @SimonCropp )